### PR TITLE
sftp: add support for ~/.ssh/config, fixes #37

### LINKED ISSUE #37: SFTP backend does not use local SSH-client's config
Coming from duplicity - which AFAIK also uses paramiko for accessing SFTP endpoints - I expected to be able to log into the SFTP backend the same (or similar) way.

For duplicity using the SFTP backend my URL is rather simple: `sftp://bck-host//storage/backup/duplicity`.

User, Port, IdentityKeyFile, etc. are supplemented from respective host defined in `~/.ssh/config`.

Example ssh-config-file:

```
Host bck-host
	IdentitiesOnly yes
	Hostname %h
	User bck
	IdentityFile ~/.ssh/id_ed25519_bck
	HashKnownHosts no
	Ciphers chacha20-poly1305@openssh.com
```

While it would be nice for borgstore to also use the local SSH-client's config (I figure paramiko already provides functionality for that), I'd at least like to have the possibility / to know how to pass an IdentityFile for login to the SFTP-server.

Is there (already) a way for doing so?

### PR DESCRIPTION
**Pull request by ThomasWaldmann**

### CODE REVIEW

**Comment by mirko** [comment]
Works for me, thanks a lot!
Some nit-picking:
 - there could be multiple `IdentityKey`s specified. From paramiko 1.9.1 onwards `paramiko.SSHConfig()` returns `identitykey` as list, no matter how many (incl. only 1) - before (<=1.9.0) it only ever returned it as string (not supporting multiple). Given that you don't specify the required version of paramiko, you might want to either do so, or handle both cases (`if not isinstance(cfg["identityfile"], list): cfg["identityfile"] = [cfg["identityfile"]]` or alike.
 - var name `scd` - what does the `d` stand for? :) `scd` immediately triggers me in regard of the infamous smartcard-daemon, which slightly confused me at first
 - you might still want to also parse `/etc/ssh/ssh_client` instead of only `~/.ssh/config` and ~~override~~/supplement options from the former with ones from the latter. Keep in mind, that lists (e.g. `identityfile` for paramiko >= 1.9.1). Question  popping up: does ~/.ssh/config extend or override options from /etc/ssh/config which are allowed to be specified multiple times?

**Comment by mirko** [comment]
Oh, and maybe also still parse respective parts from the URL provided.

I think it's more than reasonable to assume that, if I specify `sftp://USER_X@HOST:55555/` it's connecting to port 55555 as USER_X - no matter what's defined in `~/.ssh/config` in section `HOST`.

So:
 - `/etc/ssh/ssh_config` [*]
 - `~/.ssh/config`
 - elements specified as part of the URL

_Of course_ there's also `/etc/ssh/ssh_config.d/*` nowadays. Just for the record.

**Comment by ThomasWaldmann** [comment]
I didn't find any docs about how to parse multiple ssh configs (like user and system, and includes), so I just called parse multiple times: first user, then system.

Guess we might need another PR to improve that.

**Comment by ThomasWaldmann** [comment]
I'll squash everything together after review is finished, so we can have a 0.0.3 release soon.

**Comment by mirko** [comment]
I think we could do it as simple as duplicity does it: https://git.launchpad.net/duplicity/tree/duplicity/backends/ssh_paramiko_backend.py#n210

Parse system config and put it into a dict.
Parse user config and update the dict.
Parse URL and update the dict.

Yes, that would replace instead of adding options to keys which are allowed to be specified multiple times, but I think that's more predictable and easier to do document behaviour as well.

File: src/borgstore/backends/sftp.py [hunk 1]
```diff
@@ -40,7 +40,20 @@ def __init__(self, hostname: str, path: str, port: int = 22, username: Optional[
...
+        config_path = Path("~/.ssh/config").expanduser()
+        try:
+            sc = paramiko.SSHConfig.from_path(str(config_path))
+        except FileNotFoundError:
+            scd = paramiko.SSHConfigDict()  # empty dict
+        else:
+            scd = sc.lookup(self.hostname)
+        ssh.connect(
+            hostname=scd.get("hostname", self.hostname),
+            username=scd.get("user", self.username),
```
Comment (mirko) [inline src/borgstore/backends/sftp.py:52]: Also take username from URL with precedence if provided

Comment (ThomasWaldmann) [reply src/borgstore/backends/sftp.py]: Somehow my rebase onto current master has gone very wrong and lost 4 commits.

But I still have the stuff on my other dev machine, so I'll force push that soon.

Comment (ThomasWaldmann) [reply src/borgstore/backends/sftp.py]: OK, fixed that. Just squashed all past commits together.

**Comment by ThomasWaldmann** [comment]
OK, I'll check the duplicity way now. I first parsed user then system due to the documented "first match counts" behavior when dealing with the whole configs.

**Comment by mirko** [comment]
Not tested, but looking through the changeset it's how I'd expect it to be(have)

**Comment by ThomasWaldmann** [comment]
I did some sftp tests (with the test suite, but it only works with my keys), works.

**Comment by mirko** [comment]
I can test with my setup and assumptions in a few hours if that provides any benefits

**Comment by ThomasWaldmann** [comment]
@mirko Now testing sftp is much easier, see latest commit.

**Comment by mirko** [comment]
Didn't use the test suite - but tested all possible and impossible combinations.

What was to be expected: `IdentityFile` specified in `/etc/ssh/config` gets overridden and not supplemented by / if also specified in `~/.ssh/config`.

Everything else set / replaced as intended.

**Pull request merged**